adal/src/main/java/com/microsoft/aad/adal/ADALError.java

```diff
@@ -638,6 +634,38 @@ public String getLocalizedDescription(Context context) {
             return resources.getString(resources.getIdentifier(this.name(), "string",
                     context.getPackageName()));
         }
-        return mCommonADALError.getDescription();
+        return mDescription;
+    }
+    private static final Map<String, ADALError> exceptionMap = new HashMap<String,ADALError>();
+
+    static {
+        exceptionMap.put(CommonCoreException.ANDROIDKEYSTORE_FAILED, ADALError.ANDROIDKEYSTORE_FAILED);
+        exceptionMap.put(CommonCoreException.AUTHORITY_URL_NOT_VALID, ADALError.DEVELOPER_AUTHORITY_IS_NOT_VALID_INSTANCE);
+        exceptionMap.put(CommonCoreException.AUTHORITY_VALIDATION_NOT_SUPPORTED, ADALError.DEVELOPER_AUTHORITY_CAN_NOT_BE_VALIDED);
+        exceptionMap.put(CommonCoreException.DECRYPTION_ERROR, ADALError.DECRYPTION_FAILED);
+        exceptionMap.put(CommonCoreException.DEVICE_NETWORK_NOT_AVAILABLE, ADALError.DEVICE_CONNECTION_IS_NOT_AVAILABLE);
+        exceptionMap.put(CommonCoreException.ENCRYPTION_ERROR, ADALError.ENCRYPTION_ERROR);
+        exceptionMap.put(CommonCoreException.INVALID_JWT, ADALError.JSON_PARSE_ERROR);
+        exceptionMap.put(CommonCoreException.IO_ERROR, ADALError.IO_EXCEPTION);
+        exceptionMap.put(CommonCoreException.JSON_PARSE_FAILURE, ADALError.JSON_PARSE_ERROR);
+        // TODO: :O
+        exceptionMap.put(CommonCoreException.MALFORMED_URL, ADALError.DEVELOPER_AUTHORITY_IS_NOT_VALID_URL);
+        exceptionMap.put(CommonCoreException.MULTIPLE_MATCHING_TOKENS_DETECTED, ADALError.AUTH_FAILED_USER_MISMATCH);
+        exceptionMap.put(CommonCoreException.USER_MISMATCH, ADALError.AUTH_FAILED_USER_MISMATCH);
+        exceptionMap.put(CommonCoreException.NO_SUCH_ALGORITHM, ADALError.DEVICE_NO_SUCH_ALGORITHM);
+        exceptionMap.put(CommonCoreException.UNSUPPORTED_ENCODING, ADALError.ENCODING_IS_NOT_SUPPORTED);
+
+        // The following mapping need a lot more brushing up
+        exceptionMap.put(CommonCoreServiceException.ACCESS_DENIED, ADALError.AUTH_REFRESH_FAILED_PROMPT_NOT_ALLOWED);
+        exceptionMap.put(CommonCoreServiceException.INVALID_INSTANCE, ADALError.AUTH_REFRESH_FAILED_PROMPT_NOT_ALLOWED);
+        exceptionMap.put(CommonCoreServiceException.INVALID_REQUEST, ADALError.SERVER_INVALID_REQUEST);
+        exceptionMap.put(CommonCoreServiceException.REQUEST_TIMEOUT, ADALError.SERVER_ERROR);
+        exceptionMap.put(CommonCoreServiceException.SERVICE_NOT_AVAILABLE, ADALError.SERVER_ERROR);
+        exceptionMap.put(CommonCoreServiceException.UNKNOWN_ERROR, ADALError.SERVER_ERROR);
+    }
+
+    public static ADALError fromCommon(final CommonCoreBaseException exception) {
+        // What do we do if errorCode is null?
+        return exceptionMap.get(exception.getErrorCode());
```

Comment:
Other concern is how will ADAL consume the CommonCoreServiceException error code description. As CommonCoreServiceException errors have a more refined categorization. Will this mapping miss the common core error descriptions? 

For example, how do we distinguish the errors between `CommonCoreServiceException.REQUEST_TIMEOUT`, `CommonCoreServiceException.SERVICE_NOT_AVAILABLE` and `CommonCoreServiceException.UNKNOWN_ERROR` when they are all mapped into `ADALError.SERVER_ERROR`? Will the wrapped AuthenticationException lose this part of info?

Reply:
Can we create new `ADALError` to support this case? I don't want us to be digging-in deeper for an API we're not particularly happy with, but I don't know how else we can maintain the specificity _and_ work within the constraint of having to shoehorn whatever error `common` emits into a limited set of `enums`